### PR TITLE
Fixing button state logic for payment gateways

### LINKED ISSUE #7184: Finish Setup state not working correctly with payment gateway suggestions feature
With the `payment-gateway-suggestions` feature enabled, the "Finish Setup" state (pictured below) is not displaying as expected. 

image

~~May need some clarity on exactly how we want this to work in the payments task, and exactly how different statuses play a part; ie `enabled` and `configured`, which are used somewhat interchangeably at the moment.~~

In summary, we need to change the "Finish setup" state to be triggered by installation, not enabling/configuring. Also make any changes necessary to match the following states:

**Just suggested (not installed/enabled/configured)**
Currently this shows in the "Additional Gateways" list with "Setup" button.
No change needed.

**Installed, not enabled/configured**
Currently still the same as above, but this should trigger the "Finish Setup" state.

**Installed & Enabled but not configured**
This should still show the "Finish setup state"

**Configured but not enabled**
This should just show the "Manage" button.

### PR DESCRIPTION
**Pull request by joelclimbsthings**
Fixes woocommerce/woocommerce-admin#7184

Reworking logic around various buttons and states with the payment gateways in remote payments. As per the original issue, these are the states and expected behavior:

### Core gateways (CODS & BACS)
**Default**
If setup required, display `Setup` button. Otherwise display `Enable` button.

**Enabled**
Display `Manage` button, and list in the _Enabled gateways_ list.

**Note:** These methods do not have a "finish setup" state, even if setup is required.

### External gateways with setup
**Default**
Display with `Setup` button in _additional gateways_ list.

**Installed & not configured**
Display with _Finish Setup_ UI and `Finish Setup` button.

**Installed & configured**
Display with `Manage` button, in _enabled_ list if enabled, or _additional_ otherwise.

### Screenshots

image

### Detailed test instructions:

-   Ensure remote payments is enabled, and this branch is in use
- Navigate to payment task
-   Run through the enabling/setup process with COD and BACS, and ensure the button states are as expected from description above.
- Test an external gateway that requires setup, such as Razorpay. You have to use the modified version (see [PR list](woocommerce/woocommerce#32258)) for the correct behavior.
- Test WCPay card states, which should mirror the above, but with a _Get started_ button as the default state. You'll also need to use our [modified branch](https://github.com/Automattic/woocommerce-payments/pull/2014).

### CODE REVIEW

File: src/Features/PaymentGatewaySuggestions/DefaultPaymentGateways.php [hunk 1]
```diff
@@ -261,6 +261,7 @@ public static function get_rules_for_selling_venues( $selling_venues ) {
 				'option_name'  => 'woocommerce_onboarding_profile',
 				'operation'    => '=',
 				'value'        => $venue,
+				'default'      => array(),
```
Comment (joelclimbsthings) [inline src/Features/PaymentGatewaySuggestions/DefaultPaymentGateways.php:264]: This is unrelated to the rest, but was throwing an error on my local

File: client/task-list/tasks/payments/PaymentGatewaySuggestions/components/PaymentAction.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
+/**
```
Comment (joelclimbsthings) [inline client/task-list/tasks/payments/PaymentGatewaySuggestions/components/PaymentAction.js:1]: I went ahead and separated this on the basic of local/remote payments. The old file is currently called `PaymentActionLocal`, which isn't great, but I anticipate it going away pretty quick.

File: src/Features/PaymentGatewaySuggestions/DefaultPaymentGateways.php [hunk 3]
```diff
@@ -114,7 +114,7 @@ public static function get_all() {
 					(object) array(
 						'type'      => 'base_location_country',
 						'value'     => 'IN',
-						'operation' => '=',
+						'operation' => '!=',
```
Comment (joelclimbsthings) [inline src/Features/PaymentGatewaySuggestions/DefaultPaymentGateways.php:117]: Noticed that somehow razorpay and paypal had incorrect visibility rules here.

**Review by joshuatf — COMMENTED** [review]
Thanks for the fixes here, @joelclimbsthings!

Left a few comments, but also noticed (not sure if directly related to this PR) that bacs setup is no longer working:

Screen Shot 2021-06-16 at 6 59 02 PM

File: client/task-list/tasks/payments/LocalPayments.js [hunk 4]
```diff
@@ -232,6 +232,9 @@ export const LocalPayments = ( { query } ) => {
 							'Try the new way to get paid. Securely accept credit and debit cards on your site. Manage transactions without leaving your WordPress dashboard. Only with WooCommerce Payments.',
 							'wc-admin'
 						),
+						enabled: wcPayMethod.isEnabled,
```
[the remark this answers is not in the document]
Comment (joelclimbsthings) [reply client/task-list/tasks/payments/LocalPayments.js:235]: It is and it isn't. The `WCPaySuggestion` component is still directing to the newer version of the `PaymentAction` component. The `enabled` flag may not ultimately be used in this case, but in order to supply the `PaymentAction` component accurate props, it's still necessary to have this here. 

Just considering this a temporary stop-gap as long as the local component exists.

File: client/task-list/tasks/payments/PaymentGatewaySuggestions/components/PaymentAction.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+					{ __( 'Enable', 'woocommerce-admin' ) }
+				</Button>
+			);
+		}
+
+		return <ManageButton />;
+	}
+
+	// This isolates core gateways that include setup
+	if ( hasSetup && ! hasPlugins ) {
```
[the remark this answers is not in the document]
Comment (joelclimbsthings) [reply client/task-list/tasks/payments/PaymentGatewaySuggestions/components/PaymentAction.js:95]: Oh yes, logic math! Good catch 👍🏻

**Comment by joelclimbsthings** [comment]
Thanks for the review @joshuatf !
>  also noticed (not sure if directly related to this PR) that bacs setup is no longer working:

Well that's weird, I can't seem to reproduce this at all. This is how it appears on my local right now:

image

I've double checked to make sure I've pushed all my changes, and we're good there. It would only display the `Enabled` button if `hasSetup` is false, which should be true when a slotfill fill exists. Is the fill not being picked up in [this block](https://github.com/woocommerce/woocommerce-admin/blob/4451155278e6f8c49cd6791f9e618b31a60041c6/client/task-list/tasks/payments/PaymentGatewaySuggestions/components/List/Item.js#L43) for some reason?

**Review by joshuatf — APPROVED** [review]
This is testing well on my end and LGTM!  :shipit: 

The BACS issue was related to my Gutenberg install not registering the plugin properly and unrelated to this issue.

**Pull request merged**